docs/design/mono/embedding-api-deprecation.md

```diff
@@ -0,0 +1,68 @@
+
+## API stability
+
+API stability is more complicated. There are two distinct cases here: .NET 6 and the behavior moving forward.
+
+With the .NET 6 release specifically, we have an opportunity to do some initial API cleanup work because relatively few desktop users have migrated over to .NET 5 Mono. There is some low-hanging fruit to potentially clean up immediately, before we have significant migration: irrelevant functionality like `mono_domain_set` or `refonly` parameters, legacy compatibility hacks like `MonoLookupPInvokeStatus`, and subtly-incorrect behavior like fixed-size GC handles. 
+
+The .NET 6 immediate removal should only be used for functions that will allow for significant cleanup of the embedding API, runtime internals, or that would otherwise be likely to result in difficult to diagnose customer bugs or crashes at runtime. There are a lot of functions we might not want to remove for .NET 6 for ease of migration or simply time constraints. The earlier `mono_assembly_load` example is a potential case of this, but we have many similar cases of symbol variants. We want to remove many of them eventually, but for one of the above reasons it may not happen for .NET 6.
+
+To handle this moving forward, we will use a `MONO_DEPRECATED` define, which corresponds to the `[[deprecated]]` attribute on relevant compilers. For every API marked with this, we should have an entry on a new page on the Mono site explaining the deprecation and the suggested alternative. This creates a moderate bar for deprecations, but still leaves them as a possibility where necessary.
+
+Any function appearing in a major release marked deprecated may, and likely will go away in the following major release. This means that if `mono_assembly_load` is deprecated for .NET 6, it can be removed at any time leading up to .NET 7 or beyond, including during the previews.
+
+Deprecation is the preferred path for functions likely to see heavy use with embedders in .NET 6 and the _only_ option past that point. .NET 6 is the only release where removal may occur without the attribute first being present in a major release, though this should be avoided with functions likely to see heavy use in existing embedding workloads.
+
+# Points for further discussion (to address before merging)
+
+For deprecated APIs that include useless arguments or are irrelevant for netcore, should we silently ignore them/do nothing or throw a runtime error? What would be more useful for embedders, given that the deprecations allow them to detect potentially problematic functions at compile time?
+
+How does this work with LTS releases? Should we try and deprecate on LTS and remove on non-LTS when possible, or does it not matter?
+
+Should we include in the function declaration what release a given function was added or deprecated in? This would probably take the form of replacing `MONO_API` with `SINCE` and `DEPRECATED` calls, but increases complexity and effort for potentially limited value.
```

Comment:
I don't know about replacing `MONO_API` by "deprecated" and "since" but I think it would be good to have that information available in the headers (as opposed to having to hunt it down from git). (I don't think we need to do annotate every existing symbol, but as we deprecate/add symbols, it would be nice to note the release)

Reply:
I personally think we should either enforce it via the defines or it may not be applied consistently. You'd use `MONO_API` for all the existing functions, but anything new uses `SINCE` or `DEPRECATED` instead, and that forces people to be conscious of the change. I don't feel that strongly about it, however.

Reply:
Upon reflection, I think having a mix in public headers would probably not be great. I think it would be better to replace existing `MONO_API` calls with `SINCE(net5)` and then use the new defines going forward. `MONO_API` would be fine for exported functions not present in a public header.

Reply:
I guess I'm unclear why we need to change `MONO_API`.
Can't we have something like this:
```c
MONO_RT_EXTERNAL_ONLY
MONO_API
MONO_API_DEPRECATED(net6, "safe to remove callers")
void mono_domain_set (MonoDomain *ignored);

MONO_API
MONO_RT_EXTERNAL_ONLY
MONO_API_SINCE(net6)
void mono_shiny_new_thing (void);
```

Which are defined as
```c
#define MONO_API_DEPRECATED(dep_release, replacement_msg)` \
  __attribute__((deprecated(replacement_msg ", the function will be removed in the release following " #dep_release))

#define MONO_API_SINCE(since) /* empty */
```